### PR TITLE
purecap kernel MD support routines

### PR DESCRIPTION
**Pull request by qwattash**
Implement assembly support routines for MIPS and RISCV

### CODE REVIEW

File: sys/mips/cheri/copystr_c.S [hunk 1]
```diff
@@ -75,12 +73,30 @@ cstr_return_toolong:
 	li		v0, ENAMETOOLONG
 	daddiu		t1, t1, -1	/* Count is 1 too high */
 cstr_return_nul:	/* Store length read if c5 is non-NULL */
+#ifdef __CHERI_PURE_CAPABILITY__
+        cbez            $c5, cstr_return
```
Comment (brooksdavis) [inline sys/mips/cheri/copystr_c.S:77]:
```suggestion
	cbez            $c5, cstr_return
```

File: sys/riscv/riscv/support.S [hunk 2]
```diff
@@ -40,15 +40,51 @@ __FBSDID("$FreeBSD$");
 
 #include "assym.inc"
 
+/*
+ * Hybrid kernels use .cap loads and stores whereas purecap kernels use
+ * capmode loads and stores.
+ */
```
Comment (brooksdavis) [inline sys/riscv/riscv/support.S:46]: Should these and the ones in copyout.S move to a .inc or .h file?

Comment (bsdjhb) [reply sys/riscv/riscv/support.S:46]: They aren't generically useful (at least, you wouldn't use them in userland I think) and their names are quite generic, so I wouldn't want to put them in something like <machine/asm.h>.  I didn't end up using them anywhere else (and the two files end up using different sets, though copyinout's set is just a subset of these).

**Review by bsdjhb — COMMENTED** [review]
The order of commits in the history is kind of odd because the second support.S commit references the copyinout commit, but the copyinout commit then comes third.  I think in the original branch the RISC-V commits were probably ordered as '2 4 3'.  I think either the order should be restored, or the two support.S commits should be collapsed into a single commit (using the log from the first commit).

File: sys/mips/cheri/support.S [hunk 3]
```diff
@@ -0,0 +1,831 @@
...
+	clc	$c1, zero, PC_CURPCB($c1)
+	CAPTABLE_LOAD($c2, copyerr)	# prepare the onfault jump point
+	CAPCALL_LOAD($c12, copystr_c)
+	cjalr	$c12, $c17
+	csc	$c2, zero, U_PCB_ONFAULT($c1)
+
+	GET_CPU_PCPU($c1)
+	clc	$c1, zero, PC_CURPCB($c1)
+	CHERI_NULL($c2)
+	csc	$c2, zero, U_PCB_ONFAULT($c1)
```
Comment (arichardson) [inline sys/mips/cheri/support.S:133]:
```suggestion
	csc	$cnull, zero, U_PCB_ONFAULT($c1)
```
Probably not much point wasting time improving MIPS assembly, but I just saw this macro seems unnecessary.

Comment (qwattash) [reply sys/mips/cheri/support.S:133]: Yes, I attempted to remove all uses of `CHERI_NULL`, I think it might have been in another PR.

**Pull request merged**